services/QuillLMS/app/models/unit_activity.rb

```diff
@@ -82,6 +84,18 @@ def from_valid_date_for_activity_analysis?
     end
   end
 
+  def adjust_due_date_for_timezone
+    return unless due_date.present? && unit&.user&.time_zone&.present?
+
+    self.due_date = due_date - unit.user.utc_offset
+  end
+
+  def adjust_publish_date_for_timezone
+    return unless publish_date.present? && unit&.user&.time_zone&.present?
+
+    self.publish_date = publish_date - unit.user.utc_offset
+  end
```

Comment:
This code is making almost the inverse assumption - that the passed value is coming in UTC (that's why it's possible to subtract the offset and get the equivalent of the UTC time in the user's timezone). The frontend is set up to only ever pass utc objects back (all conversions are in `moment.utc`, which handles that - we can't rely on the browser time because it doesn't necessarily match the user's saved timezone on the backend). I think you're right that we shouldn't rely on the frontend and should instead confirm that the time is in utc before the subtraction. I'll update it to reflect that.